c-api/compat-5.3.c

```diff
@@ -14,6 +15,61 @@
 /* definitions for Lua 5.1 only */
 #if defined(LUA_VERSION_NUM) && LUA_VERSION_NUM == 501
 
+#ifndef COMPAT53_LUA_FILE_BUFFER_SIZE
+#define COMPAT53_LUA_FILE_BUFFER_SIZE 4096
+#endif // Lua File Buffer Size
+
+#define compat53_f_parser_buffer COMPAT53_CONCAT(COMPAT53_PREFIX, compat53_f_parser_buffer)
+struct compat53_f_parser_buffer {
+  FILE* file;
+  size_t start;
+  char buffer[COMPAT53_LUA_FILE_BUFFER_SIZE];
+};
+
+#if defined(__GNUC__) && (__GNUC__ < 6)  && defined(__MINGW32__)
+// MinGW 5.x.x and below
+// So as #485 indicates, MinGW doesn't have this function of 5.3 of MinGW
+// time to cover for C's standard library. again
+static int strerror_r(int errnum, char *buf, size_t len) {
+	// did we get a crap buffer?
+	if (buf == NULL) {
+		// considered a range error
+		// write errno, then return
+		errno = ERANGE;
+		return ERANGE;
+	}
+	//
+	if ((errnum < 0) || (errnum >= sys_nerr)) {
+		// failure
+		snprintf(buf, len, "unknown error: %d", errnum);
+		// write errno, then return
+		errno = EINVAL;
+		return EINVAL;
+	}
+	// this is technically thread unsafe if `strerror` doesn't use threadlocal storage
+	// 
+	// pthreads isn't guaranteed either
+	// Now. The stdc lib devs probably use thread_local on their stuff
+	// but just in case, we'll do this here to be super safe
+	// this should protect us from data in-fighting across threads
+#if defined(_MSC_VER) defined(__BORLANDC__)
+	__declspec(thread) 
+#else
+	__thread 
+#endif // VC++ vs.everyone else
+	char* hopeitsthreadlocal = strerror(errnum);
```

Comment:
strerror is not threadsafe, maybe use [sys_errlist](http://www.club.cc.cmu.edu/~cmccabe/blog_strerror.html) or strerror_s/strerror_r, as below in the PR.